com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs

```diff
@@ -0,0 +1,605 @@
+
+            // Write the new source file.
+            using (FileStream fs = outputFile.OpenWrite())
+            {
+                using (StreamWriter writer = new StreamWriter(fs))
+                {
+                    writer.AutoFlush = true;
+                    writer.Write(template);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Attempts to acquire a <see cref="FileInfo"/> object representing the requested asset.
+        /// </summary>
+        /// <param name="guid">The guid that represents the asset.</param>
+        /// <param name="fileInfo"><see cref="FileInfo"/> object representing the asset.</param>
+        /// <param name="error">Error message.</param>
+        /// <returns>True if the asset was acquired, or false.</returns>
+        private bool GetAsset(
+            string guid,
+            out FileInfo fileInfo,
+            out string error)
+        {
+            fileInfo = null;
+            error = string.Empty;
+
+            string filePath = AssetDatabase.GUIDToAssetPath(guid);
+            if (string.IsNullOrWhiteSpace(filePath))
+            {
+                error = "Unable to resolve template asset, please file an MRTK3 bug.";
+                return false;
+            }
+
+            FileInfo fi = new FileInfo(filePath);
+            if (!fi.Exists)
+            {
+                error = $"Template file {fi.Name} could not be located, please file an MRTK3 bug.";
+                return false;
+            }
+
+            fileInfo = fi;
+            return true;
+        }
+
+        /// <summary>
+        /// Replaces template tokens with customer data.
+        /// </summary>
+        /// <param name="template">The contents of the template.</param>
+        /// <returns>The updated contents of the template.</returns>
+        private string ReplaceTokens(string template)
+        {
+            template = template.Replace("%NAMESPACE%", SubsystemNamespace);
+            template = template.Replace("%SUBSYSTEMBASECLASSNAME%", BaseClassName);
+            template = template.Replace("%SUBSYSTEMNAME%", SubsystemName);
+            template = template.Replace("%CONFIGNAME%", ConfigurationName);
+            template = template.Replace("%ORGANIZATION%", OrganizationName);
+            template = template.Replace("%DISPLAYNAME%", DisplayName);
+            template = template.Replace("%RUNTIMENAME%", SubsystemNamespace.ToLower());
+
+            return template;
+        }
+
+        /// <summary>
+        /// Validates that the specified name is a valid indentifier for C#.
+        /// </summary>
+        /// <param name="name">The name to validate/</param>
+        /// <returns>True if the name is a valid C# identifier, or false.</returns>
+        private bool ValidateName(string name)
+        {
+            // Verify that the name is valid within C#
+                                        return CSharpCodeProvider.CreateProvider("C#").IsValidIdentifier(name);
```

Comment:
nit: weird indentation here

Reply:
that IS odd.... i havent seen that one 😆

Reply:
looks like something landed on my spacebar at some point... haha

Reply:
fixed